alot/commands/__init__.py

```diff
@@ -136,15 +163,16 @@ def __init__(self, mode, name, help=None, usage=None,
         self.forced = forced or {}
         self.arguments = arguments or []
 
-    def __call__(self, klass):
-        helpstring = self.help or klass.__doc__
+    def __call__(self, class_):
+        helpstring = self.help or class_.__doc__
         argparser = CommandArgumentParser(description=helpstring,
                                           usage=self.usage,
                                           prog=self.name, add_help=False)
         for args, kwargs in self.arguments:
             argparser.add_argument(*args, **kwargs)
-        COMMANDS[self.mode][self.name] = (klass, argparser, self.forced)
-        return klass
+        COMMANDS[self.mode][self.name] = (class_, argparser, self.forced)
+        REVERSE_COMMANDS[class_] = (self.name, self.mode)
```

Comment:
It looks like `mode` and `cmdname` are needed one way or another.

Hooks are accessed via `settings.get_hook(key)` where `key` is the name of the hook function. Since names of hook functions are constructed as `{pre|post}_<mode>_<cmdname>`, both `mode` and`cmdname` are needed to retrieve the hook from `settings.get_hook()`. This is why I stored `mode` and `cmdname` in the original global dictionary.

The only place where `mode` and `cmdname` are available without "storing" them first, is in the `registerCommand` decorator. However, the decorator is run on the classes *before* the settings have been loaded, so there are never any hooks to retrieve.

I pushed another solution where I store `mode` and `cmdname` as class variables, then making `prehook` and `posthook` properties of `Command` instead with only a getter.

So with this solution, no extra "lookup" structure is needed, and hooks are fetched when needed.

What do you think?